### ISSUE TITLE
docker: Add a HEALTHCHECK in Dockerfile

### ISSUE THREAD

**Issue by shouze** [minio/minio#3800]
Should be cool to add a [HEALTHCHECK](https://docs.docker.com/engine/reference/builder/#/healthcheck) in the Dockerfile ([supported since docker 1.12)

## Expected Behavior

Will help docker to knwo the health state of the container, very helpfull with [depends_on service_healthy conditions](https://docs.docker.com/compose/compose-file/#/dependson) in docker compose (introduced with version 2.1 of docker-compose file).

## Current Behavior

No healthcheck provided.

## Possible Solution

On possible solution as suggested by @harshavardhana in #3532 could be to do (we should have curl installed into the minio container):
```
HEALTHCHECK CMD curl -f http://localhost:9000/minio/index.html || exit 1
```

Another solution would be to use [madm](https://github.com/minio/minio/tree/master/pkg/madmin) would be a far more **true** healthcheck maybe?

## Context

As we use the minio docker container and we want to orchestrate the start & dependencies between our services it's very helpfull that each service can provide a healthcheck feedback to docker.

## Your Environment
* Version used: any version of minio
* Environment name and version: docker >= 1.2 with docker-compose >= 1.10
* Server type and version: Linux & macOs

**Comment by shouze** [comment]
@harshavardhana which solution would you prefer? I guess the `madm` one would be the best one no? But it would imply to pass credentials (could be done throug env vars though).

It also imply to add madm into the docker container. Tell me if I can help & how I would be very pleased 😉

**Comment by harshavardhana** [comment]
> @harshavardhana which solution would you prefer? I guess the madm one would be the best one no? But it would imply to pass credentials (could be done throug env vars though).

Yes this is possible using `madm` a simple command, we can also additionally compile `mc` https://github.com/minio/mc which already has these in place which can be used directly.

**Comment by shouze** [comment]
ok for mc, geat! I'll give a try so I guess that we don't have to even compile in fact we can download the compiled binary directly no? (https://dl.minio.io/client/mc/release/linux-amd64/mc into the Dockerfile)

**Comment by harshavardhana** [comment]
> ok for mc, geat! I'll give a try so I guess that we don't have to even compile in fact we can download the compiled binary directly no? (https://dl.minio.io/client/mc/release/linux-amd64/mc into the Dockerfile)

For the time being we might have to.. because `mc admin service status` is only available in the current master.

**Comment by shouze** [comment]
ok, got it

**Comment by shouze** [comment]
@harshavardhana ok I've made a try, so I've made a mc build inside the container, it worked (except the deadcode step if I remember, commented it for the poc).

Then I've done that kind of things:
```shell
mc config host add minio <ENDPOINT> <ACCESS_KEY> <SECRET_KEY> <SIGNATURE_VERSION>
mc admin service status --json minio
```

As minio keys can be passed trhough env vars (`MINIO_ACCESS_KEY` & `MINIO_SECRET_KEY`) I guess that in the Dockerfile I should replace the actual ENTRYPOINT (`minio`) by some entrypoint.sh that will looks like:

```shell
#!/usr/bin/env sh
set -e

mc config host add minio http://localhost:9000 $MINIO_ACCESS_KEY $MINIO_SECRET_KEY s3v4

exec minio
```

This way, we can have an alias up to date at runtime that `admin service status --json minio` will check in the docker `HEALTHCHECK` command. WDYT?

Also... I would be cool to have an official/tagged release of mc instead of having to compile it inside the [Dockerfile](https://github.com/minio/minio/blob/dce0345f8f8a4654d92dbb28d46a09f19db1d26d/Dockerfile) to avoid pollution inside it.

**Comment by harshavardhana** [comment]
> @harshavardhana ok I've made a try, so I've made a mc build inside the container, it worked (except the deadcode step if I remember, commented it for the poc).
> 
> Then I've done that kind of things:
> 
> mc config host add minio <ENDPOINT> <ACCESS_KEY> <SECRET_KEY> <SIGNATURE_VERSION>
> mc admin service status --json minio
> As minio keys can be passed trhough env vars (MINIO_ACCESS_KEY & MINIO_SECRET_KEY) I guess that in the Dockerfile I should replace the actual ENTRYPOINT (minio) by some entrypoint.sh that will looks like:
> 
> #!/usr/bin/env sh
> set -e
> 
> mc config host add minio http://localhost:9000 $MINIO_ACCESS_KEY $MINIO_SECRET_KEY s3v4
> 
> exec minio
> This way, we can have an alias up to date at runtime that admin service status --json minio will check in the docker HEALTHCHECK command. WDYT?

Right now we don't have a script, we have to come up with an entrypoint.sh script which creates all the necessary environment for all of this to work. 

> 
> Also... I would be cool to have an official/tagged release of mc instead of having to compile it inside the Dockerfile to avoid pollution inside it.

We can do `mc admin service status` and it would serve the purpose for this - the official release can be found here - https://hub.docker.com/r/minio/mc/ or you can download the mc binary etc.

**Comment by shouze** [comment]
> Right now we don't have a script, we have to come up with an entrypoint.sh script which creates all the necessary environment for all of this to work.

ok will do that thx for the confirmation.

> We can do mc admin service status and it would serve the purpose for this - the official release can be found here - https://hub.docker.com/r/minio/mc/ or you can download the mc binary etc.

Yup, but even if the https://hub.docker.com/r/minio/mc/ inside the docker container image is up2date, it's a bit tricky during minio container's build to extract the mc binary from the mc docker image, it's far easier to (curl) download a tagged binary that support the `mc admin service status` from the github repo.

To sum up, would be easier if you tag a new github release soon, because the latest tagged release still don't support `mc admin service status` ;)

**Comment by harshavardhana** [comment]
> To sum up, would be easier if you tag a new github release soon, because the latest tagged release still don't support mc admin service status ;)

Oh yes this is under review should happen soon. Thanks for spending time on this @shouze - you are awesome :v:

**Comment by shouze** [comment]
@harshavardhana do you think it's ok now to add a new tag on https://github.com/minio/mc/releases repo?

Then I could achieve a PR for this issue very soon. Thx.

**Comment by harshavardhana** [comment]
> @harshavardhana do you think it's ok now to add a new tag on https://github.com/minio/mc/releases repo?

We are still under testing, we will be making a release soon.

**Comment by deekoder** [comment]
Depends on working admin-cli in mc.

**Comment by harshavardhana** [comment]
This is dependent on `mc` release and can be done after.

**Comment by beniwtv** [comment]
This might be a good solution for the provided Docker images of minio, but I think this has it's issues for other environments, that might not use Docker Compose with Docker health checks.

If you do not use the Docker image, you will have to set things up manually, which may lead to problems people are not aware of - like secure storage of the credentials and such.

We use Consul and Nomad in our environment, and while it can execute scripts to do health checking - we would pretty much prefer doing a simple HTTP check, as done with most other servers.

This check should not need authentication, and doesn't have to return any sensitive data - a simple 200 OK header or 500 on error would be sufficient.

Note that "/minio/index.html" only works if the browser is enabled - which we don't have as it's not needed in our environment, and I don't know if it actually checks health besides the browser part being up.

**Comment by harshavardhana** [comment]
Fixed by #4550
